### PR TITLE
feat: Middleware to add Content-Security-Policy headers

### PR DESCRIPTION
**Pull request by timmc-edx**
The goal is to allow rapid experimentation until we have a reasonably useful CSP header for at least on large IDA, and then to move this code to a more appropriate library.

### CODE REVIEW

File: edx_arch_experiments/__init__.py [hunk 1]
```diff
@@ -2,6 +2,6 @@
 A plugin to include applications under development by the architecture team at edx .
 """
 
-__version__ = '1.1.2'
+__version__ = '1.2.0+timmc1'
```
Comment (timmc-edx) [inline edx_arch_experiments/__init__.py:5]: TODO: Remove this, used only for manual devstack testing.

File: edx_arch_experiments/csp/middleware.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    # .. setting_name: CSP_REPORTING_NAME
+    # .. setting_default: None
+    # .. setting_description: Used for CSP Level 3 reporting. This sets the name to use in the
+    #   report-to CSP field and the Reporting-Endpoints header. This should generally be an
+    #   alphanumeric string; other characters such as hyphen and underscore are also allowed.
+    #   See https://www.rfc-editor.org/rfc/rfc8941.html#section-3.3.4 for full grammar.
+    reporting_endpoint_name = getattr(settings, 'CSP_REPORTING_NAME', None)
+
+    # RFC 8941 Token is a RFC 7230 that must start with ALPHA or *, and can also contain / and :.
+    sfToken = "[a-zA-Z*][a-zA-Z0-9!#$%&'*+\\-.^_`|~\":/]*"
```
Comment (timmc-edx) [inline edx_arch_experiments/csp/middleware.py:57]: Author's note: Not sure if I should bother with this arcane regex or if we should be stricter than the actual spec and just say "alphanumeric, hyphens, and underscore, but has to start with a letter" to make this check simpler.

**Review by rgraber — COMMENTED** [review]
Did you look for other methods we may have for adding headers somewhere? I'd be surprised to learn we don't have any. It seems like a common use case for middleware.

**Comment by timmc-edx** [comment]
Yeah, I thought for sure there would be a middleware or library (that we already use) that just attaches a configured set of generic headers to every response, but I couldn't find anything! I did consider adding my own, but there are a couple of reasons it's nice to have something specific to CSP:

- CSP policies tend to be quite long, and being able to break them up over multiple lines is really nice. You can get YAML to strip newlines but it's pretty fragile.
- I like being able to deduplicate the reporting directives.
- The waffle flag seems useful for rollouts on large sites.

I think I'm going to go ahead and move this over to edx-django-utils, though.

**Comment by timmc-edx** [comment]
This is replaced by https://github.com/openedx/edx-django-utils/pull/289

**Pull request closed without merging**